### PR TITLE
Add support for brushless hardware

### PR DESCRIPTION
**Pull request by jedahan**
Should be ready to merge

* ~~check real world dimensions are correct~~
* ~~fix final pen position after a cancel~~
* ~~get the tests to pass~~

### CODE REVIEW

File: src/planning.ts [hunk 1]
```diff
@@ -476,109 +520,115 @@ function dedupPoints(points: Vec2[], epsilon: number): Vec2[] {
...
-        blocks.push(b);
+        blocks.push(b)
       }
-    });
-  });
-  return new XYMotion(blocks);
+    })
+  })
+  return new XYMotion(blocks)
 }
```
Comment (jedahan) [inline src/planning.ts:602]: Semicolon noise snuck in, I'm hoping after this merge to do `npm run format` and have that be a separate PR

File: src/planning.ts [hunk 2]
```diff
@@ -476,109 +520,115 @@ function dedupPoints(points: Vec2[], epsilon: number): Vec2[] {
...
+  })
+
+  // Move to {x: 0, y: 0}
+  motions.push(constantAccelerationPlan([curPos, { x: 0, y: 0 }], profile.penUpProfile))
+  // TODO: should minPenPosition be
+  //      a) profile.penUpPosition?
+  //      b) profile.penDownPosition?
+  //      c) device.penPctToPos(0)
+  //      d) device.penPctToPos(100)
+  //      e) device.penPctToPos(profile.penUpPos < profile.penDownPos ? 100: 0)
```
Comment (jedahan) [inline src/planning.ts:632]: Can remove this, it looks like it goes back to pen up position correctly on brushless, but someone should check on regular.

**Comment by jedahan** [comment]
Plan tests fail, gonna have to take another look at it tomorrow unless you have a chance to look at it.

File: src/planning.ts (hunk 2 above)
Comment (alexrudd2) [reply src/planning.ts]: I think I need to write up an issue that captures all the penUp/penDown positioning oddities.

**Comment by alexrudd2** [comment]
It runs pretty well on a RPi 4!

We did notice that the default settings for brushless are pretty slow.  It looks like you used the defaults from `AxiDraw` and not `AxidrawFast`.  This isn't a problem (they can be changed after all), just wanted to make sure you realized.

Got an OOM error that I haven't seen before.  No idea if this is new or was triggered by a longer/slower plot (https://github.com/alexrudd2/saxi/issues/101).

`npm run dev` didn't work for me; I used `npx .` or `npm install -g . && saxi`.  Did not investigate further.

Did not try the UI brushless checkbox; does it work for you?

**Comment by alexrudd2** [comment]
> Did not try the UI brushless checkbox; does it work for you?

It works!

**Comment by jedahan** [comment]
I'll update the speed values and look at the OOM.

**Comment by jedahan** [comment]
Its not obvious to me what parameters need to be updated to speed up the motor - all the past brushless diffs use the same settings. Mind proposing what changes to fix things? (I assume its accel and maxVeloctiy).

I started looking at https://github.com/evil-mad/axidraw/blob/master/inkscape%20driver/axidraw_conf.py and I think it'll take some time to document what params are in what units, and how they map to each other.

Theres also a little confusion of PlanOptions vs Device vs ToolingProfile. Might be reading the wrong fields in some of the code.

**Comment by jedahan** [comment]
Some notes regarding the planning test - this is stream of consciousness.

What motions do we expect for a single point at 10, 10? `plan([[{x: 10, y: 10}]], AxidrawFast)`

Logically I would expect:
```js
up = profile.penUpPos, down = profile.penDownPos

PenMotion(???unknownStartingPosition???,  up)
XYMotion([0,0], [10,10])
PenMotion(up -> down)
PenMotion(down -> up)
XYMotion([10,10], [0, 0])
```
What do we get if we print the plan?

```json
  [
      { xy: [ { x: 0, y: 0 }, { x: 10, y: 10 } ] }, // move from 0,0 to 10,10
      { z: [ 17750, 15700 ] }, // drop the pen from up to down
      { xy: [ { x: 10, y: 10 }, { x: 10, y: 10 } ] }, // move from 10,10 to 10,10
      { z: [ 15700, 17750 ] }, // lift the pen
      { xy: [ { x: 10, y: 10 }, { x: 0, y: 0 } ] } // move from 10,10, to 0, 0
    ]
```

Can we query the initial position? Should be able to. Might be a TODO in the `plan()` function in this PR. 

Should we always start with a PenMotion to up?

Why do we have an XYMotion with start === end?

**Comment by alexrudd2** [comment]
> Its not obvious to me what parameters need to be updated to speed up the motor - all the past brushless diffs use the same settings. Mind proposing what changes to fix things? (I assume its accel and maxVeloctiy).

I can't figure it out either, since (as you said) the settings are the same - and the brushed motor is unaffected.  

Perhaps `AxidrawBrushless.stepsPerMm` isn't correct for the servo?  Do your plots have the correct real-world dimensions?

File: src/__tests__/planning.test.ts [hunk 3]
```diff
@@ -25,8 +30,8 @@ describe("plan", () => {
 
     expect(xyMotions(p)).toEqual([
       {from: {x: 0, y: 0}, to: {x: 10, y: 10}, penPos: 0},
-      {from: {x: 10, y: 10}, to: {x: 10, y: 10}, penPos: AxidrawFast.penDownPos},
-      {from: {x: 10, y: 10}, to: {x: 0, y: 0}, penPos: device.penPctToPos(0)},
+      {from: {x: 10, y: 10}, to: {x: 10, y: 10}, penPos: positions.up},
```
Comment (alexrudd2) [inline src/__tests__/planning.test.ts:33]: I think this is a typo, because `penDownPos` became `positions.up` instead of `positions.down`?

File: src/planning.ts [hunk 4]
```diff
@@ -336,7 +336,7 @@ export class Plan {
         // TODO: Remove this hack by storing the pen-up/pen-down heights
         // in a single place, and reference them from the PenMotions.
         if (j === this.motions.length - 1) {
-          return new PenMotion(this.minPenPosition, penUpHeight, motion.duration())
+          return new PenMotion(Axidraw.penPctToPos(0), penUpHeight, motion.duration())
```
Comment (alexrudd2) [inline src/planning.ts:339]: We don't want to hard-code this, but I wanted to make it explicit where the logic change had occurred.

Comment (jedahan) [reply src/planning.ts:339]: oh! good catch, yeah this should be fixed. Honestly I have not looked into why PenMotions need a starting position in their motion definition - we should be able to know what the current position is since we only ever configure 2 pen positions - up and down - and use that to calculate the target duration delay.

**Comment by jedahan** [comment]
Updated the main comment to include checking that real world dimensions are correct.

**Comment by alexrudd2** [comment]
> Updated the main comment to include checking that real world dimensions are correct.

Probably relevant: https://github.com/nornagon/saxi/issues/120

**Comment by jedahan** [comment]
Made some special rectangles, all came out exactly to the `mm` :)

**Comment by alexrudd2** [comment]
> Made some special rectangles, all came out exactly to the `mm` :)

Congrats and thank you, that's great news!

**Comment by jedahan** [comment]
So I tried a more complex svg, and now the stuttering issues are back. also cancel is unresponsive. had to kill the process.

Cannot check if this is a problem upstream or with this patchset since I only have one axidraw.

So we might need to remove the webserial support after all for it to work.

**Comment by alexrudd2** [comment]
> So I tried a more complex svg, and now the stuttering issues are back. also cancel is unresponsive. had to kill the process.
> 
> Cannot check if this is a problem upstream or with this patchset since I only have one axidraw.
> 
> So we might need to remove the webserial support after all for it to work.

Pretty sure it's upstream.  And intermittent.  :S  You can usually see in the console there are lots of connects/disconnects.

**Comment by jedahan** [comment]
So I don't see lots of connects/disconnects in the web console, but I was able to reproduce the issue even after restarting saxi with this svg.
 
9662b7d7-22ba-4e87-b23d-b8422e0c964a

I think this PR is ready, and once its merged I'll start on some more incremental refactoring / formatting PRs.

**Comment by alexrudd2** [comment]
> So I don't see lots of connects/disconnects in the web console, but I was able to reproduce the issue even after restarting saxi with this svg.


I can't reproduce with that SVG (on an M1 Macbook Air).  That said, I don't see a difference between this branch and main (with a brushed motor).  I think the underlying issue is present upstream.  See how there are 5 redundant websocket (NOT webserial) messages at once.  I've seen this for months and never been able to figure it out
Screenshot 2023-10-14 at 2 38 39 PM

**Comment by jedahan** [comment]
Did a little digging by printing out the number of clients when connecting on the server, and a single tab was showing 10 connections. Seems an issue.
